### PR TITLE
Improvements in neural network training and evaluation

### PR DESCRIPTION
**Pull request by HenKlei**
This PR slightly refactors the neural network training loop to avoid unnecessary forward passes. Further, the evaluation of the networks is improved in terms of computational efficiency by avoiding unnecessary gradient information assembly during forward pass.

### CODE REVIEW

**Review by ullmannsven — APPROVED** [review]
looks good, no comments from my end!

**Comment by HenKlei** [comment]
Do you want to take a look @sdrave or are you fine with merging?

**Review by sdrave — APPROVED** [review]
